src/packet/gre.rs.in

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2016 Robert Collins <robertc@robertcollins.net>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#[cfg(test)]
+use packet::Packet;
+
+use packet::PrimitiveValues;
+
+use pnet_macros_support::types::*;
+
+/// GRE Packet.
+///
+/// Note that routing information from RFC 1701 is not implemented, packets
+/// with routing_present true will have incorrect payload lengths.
+#[packet]
+pub struct Gre {
+    checksum_present: u1,
+    routing_present: u1,
+    key_present: u1,
+    sequence_present: u1,
+    strict_source_route: u1,
+    recursion_control: u3,
+    zero_flags: u5,
+    version: u3,
+    protocol_type: u16be, // 0x800 for ipv4 [basically an ethertype
+    #[length_fn = "gre_checksum_length"]
+    checksum: Vec<U16BE>,
+    #[length_fn = "gre_offset_length"]
+    offset: Vec<U16BE>,
+    #[length_fn = "gre_key_length"]
+    key: Vec<U32BE>,
+    #[length_fn = "gre_sequence_length"]
+    sequence: Vec<U32BE>,
+    // XXX: Routing would go here.
+    #[payload]
+    payload: Vec<u8>,
+}
+
+fn gre_checksum_length(gre: &GrePacket) -> usize {
+    (gre.get_checksum_present() | gre.get_routing_present()) as usize * 2
+}
+
+fn gre_offset_length(gre: &GrePacket) -> usize {
+    (gre.get_checksum_present() | gre.get_routing_present()) as usize * 2
+}
+
+fn gre_key_length(gre: &GrePacket) -> usize {
+    gre.get_key_present() as usize * 4
+}
+
+fn gre_sequence_length(gre: &GrePacket) -> usize {
+    gre.get_sequence_present() as usize * 4
+}
+
+
+/// u16be, but we can't use that directly in a Vec :(
+#[packet]
+pub struct U16BE {
+    number: u16be,
+    #[length_fn = "u16be_payload_length"]
+    #[payload]
+    unused: Vec<u8>,
+}
+
+fn u16be_payload_length(_: &U16BEPacket) -> usize {
+    0
+}
+
+
+/// u32be, but we can't use that directly in a Vec :(
+#[packet]
+pub struct U32BE {
+    number: u32be,
+    #[length_fn = "u32be_payload_length"]
+    #[payload]
+    unused: Vec<u8>,
+}
+
+fn u32be_payload_length(_: &U32BEPacket) -> usize {
+    0
+}
+
+#[test]
+fn gre_packet_test() {
+    let mut packet = [0u8; 4];
+    {
+        let mut gre_packet = MutableGrePacket::new(&mut packet[..]).unwrap();
+        gre_packet.set_protocol_type(0x0800);
+        assert_eq!(gre_packet.payload().len(), 0);
+    }
+
+    // TODO: get traces of gre packets with checksums, sequence numbers etc
+    // 0x00, 0x00, 0x08, 0x00, - no checksum, no sequence, protocol ipv4
```

Comment:
Good to know - thought I should ask!

I'm usually fairly lax about tests for packet types now, since the code is auto-generated, so it should "just work". GRE is pushing the limits of what pnet_macros can do though, so I'd appreciated the test cases to be sure it's working.